File: Raven.Abstractions/Json/Linq/RavenJObject.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using Raven.Abstractions.Extensions;
using Raven.Abstractions.Json;
using Raven.Imports.Newtonsoft.Json;
using Raven.Imports.Newtonsoft.Json.Linq;
using Raven.Json.Utilities;

namespace Raven.Json.Linq
{
	/// <summary>
	/// Represents a JSON object.
	/// </summary>
	public class RavenJObject : RavenJToken, IEnumerable<KeyValuePair<string, RavenJToken>>
	{
		private readonly IEqualityComparer<string> comparer;

		/// <summary>
		/// Gets the node type for this <see cref="RavenJToken"/>.
		/// </summary>
		/// <value>The type.</value>
		public override JTokenType Type
		{
			get { return JTokenType.Object; }
		}

		internal DictionaryWithParentSnapshot Properties { get; set; }

		public int Count
		{
			get { return Properties.Count; }
		}

		public ICollection<string> Keys
		{
			get { return Properties.Keys; }
		}

		public RavenJObject WithCaseInsensitivePropertyNames()
		{
			var props = new DictionaryWithParentSnapshot(StringComparer.InvariantCultureIgnoreCase);
			foreach (var property in Properties)
			{
				props[property.Key] = property.Value;
			}
			return new RavenJObject(props);
		}

		/// <summary>
		/// Initializes a new instance of the <see cref="RavenJObject"/> class.
		/// </summary>
		public RavenJObject() :this(StringComparer.InvariantCulture)
		{
		}

		public RavenJObject(IEqualityComparer<string> comparer)
		{
			this.comparer = comparer;
			Properties = new DictionaryWithParentSnapshot(comparer);
		}

		public RavenJObject(RavenJObject other)
		{
			Properties = new DictionaryWithParentSnapshot(other.comparer);
			foreach (var kv in other.Properties)
			{
				Properties.Add(kv);
			}
		}

		private RavenJObject(DictionaryWithParentSnapshot snapshot)
		{
			Properties = snapshot;
		}

		internal override bool DeepEquals(RavenJToken other)
		{
			var t = other as RavenJObject;
			if (t == null)
				return false;

			return base.DeepEquals(other);
		}

		/// <summary>
		/// Gets the <see cref="RavenJToken"/> with the specified key converted to the specified type.
		/// </summary>
		/// <typeparam name="T">The type to convert the token to.</typeparam>
		/// <param name="key">The token key.</param>
		/// <returns>The converted token value.</returns>
		public override T Value<T>(string key)
		{
			return this[key].Convert<T>();
		}

		/// <summary>
		/// Gets or sets the <see cref="RavenJToken"/> with the specified property name.
		/// </summary>
		/// <value></value>
		public RavenJToken this[string propertyName]
		{
			get
			{
				RavenJToken ret;
				Properties.TryGetValue(propertyName, out ret);
				return ret;
			}
			set { Properties[propertyName] = value; }
		}

		public override RavenJToken CloneToken()
		{
			return CloneTokenImpl(new RavenJObject());
		}

		internal override IEnumerable<KeyValuePair<string, RavenJToken>> GetCloningEnumerator()
		{
			return Properties;
		}

		/// <summary>
		/// Creates a <see cref="RavenJObject"/> from an object.
		/// </summary>
		/// <param name="o">The object that will be used to create <see cref="RavenJObject"/>.</param>
		/// <returns>A <see cref="RavenJObject"/> with the values of the specified object</returns>
		public static new RavenJObject FromObject(object o)
		{
			return FromObject(o, JsonExtensions.CreateDefaultJsonSerializer());
		}

		/// <summary>
		/// Creates a <see cref="RavenJArray"/> from an object.
		/// </summary>
		/// <param name="o">The object that will be used to create <see cref="RavenJArray"/>.</param>
		/// <param name="jsonSerializer">The <see cref="JsonSerializer"/> that will be used to read the object.</param>
		/// <returns>A <see cref="RavenJArray"/> with the values of the specified object</returns>
		public static new RavenJObject FromObject(object o, JsonSerializer jsonSerializer)
		{
			RavenJToken token = FromObjectInternal(o, jsonSerializer);

			if (token != null && token.Type != JTokenType.Object)
				throw new ArgumentException("Object serialized to {0}. RavenJObject instance expected.".FormatWith(CultureInfo.InvariantCulture, token.Type));

			return (RavenJObject)token;
		}

		/// <summary>
		/// Loads an <see cref="RavenJObject"/> from a <see cref="JsonReader"/>. 
		/// </summary>
		/// <param name="reader">A <see cref="JsonReader"/> that will be read for the content of the <see cref="RavenJObject"/>.</param>
		/// <returns>A <see cref="RavenJObject"/> that contains the JSON that was read from the specified <see cref="JsonReader"/>.</returns>
		public new static RavenJObject Load(JsonReader reader)
		{
			if (reader.TokenType == JsonToken.None)
			{
				if (!reader.Read())
					throw new Exception("Error reading RavenJObject from JsonReader.");
			}

			if (reader.TokenType != JsonToken.StartObject)
				throw new Exception(
					"Error reading RavenJObject from JsonReader. Current JsonReader item is not an object: {0}".FormatWith(CultureInfo.InvariantCulture, reader.TokenType));

			if (reader.Read() == false)
				throw new Exception("Unexpected end of json object");

			string propName = null;
			var o = new RavenJObject();
			do
			{
				switch (reader.TokenType)
				{
					case JsonToken.Comment:
						// ignore comments
						break;
					case JsonToken.PropertyName:
						propName = reader.Value.ToString();
						break;
					case JsonToken.EndObject:
						return o;
					case JsonToken.StartObject:
						if (!string.IsNullOrEmpty(propName))
						{
							var val = RavenJObject.Load(reader);
							o[propName] = val; // TODO: Assert when o.Properties.ContainsKey and its value != val
							propName = null;
						}
						else
						{
							throw new InvalidOperationException("The JsonReader should not be on a token of type {0}."
																	.FormatWith(CultureInfo.InvariantCulture,
																				reader.TokenType));
						}
						break;
					case JsonToken.StartArray:
						if (!string.IsNullOrEmpty(propName))
						{
							var val = RavenJArray.Load(reader);
							o[propName] = val; // TODO: Assert when o.Properties.ContainsKey and its value != val
							propName = null;
						}
						else
						{
							throw new InvalidOperationException("The JsonReader should not be on a token of type {0}."
																	.FormatWith(CultureInfo.InvariantCulture,
																				reader.TokenType));
						}
						break;
					default:
						if (!string.IsNullOrEmpty(propName))
						{
							var val = RavenJValue.Load(reader);
							o[propName] = val; // TODO: Assert when o.Properties.ContainsKey and its value != val
							propName = null;
						}
						else
						{
							throw new InvalidOperationException("The JsonReader should not be on a token of type {0}."
																	.FormatWith(CultureInfo.InvariantCulture,
																				reader.TokenType));
						}
						break;
				}
			} while (reader.Read());

			throw new Exception("Error reading RavenJObject from JsonReader.");
		}

		/// <summary>
		/// Load a <see cref="RavenJObject"/> from a string that contains JSON.
		/// </summary>
		/// <param name="json">A <see cref="String"/> that contains JSON.</param>
		/// <returns>A <see cref="RavenJObject"/> populated from the string that contains JSON.</returns>
		public new static RavenJObject Parse(string json)
		{
			try
			{
				JsonReader jsonReader = new RavenJsonTextReader(new StringReader(json));
				return Load(jsonReader);
			}
			catch (Exception e)
			{
				throw new InvalidOperationException("Could not parse json:" + Environment.NewLine + json, e);
			}
		}

		/// <summary>
		/// Writes this token to a <see cref="JsonWriter"/>.
		/// </summary>
		/// <param name="writer">A <see cref="JsonWriter"/> into which this method will write.</param>
		/// <param name="converters">A collection of <see cref="JsonConverter"/> which will be used when writing the token.</param>
		public override void WriteTo(JsonWriter writer, params JsonConverter[] converters)
		{
			writer.WriteStartObject();

			if (Properties != null)
			{
				foreach (var property in Properties)
				{
					writer.WritePropertyName(property.Key);
					if(property.Value == null)
						writer.WriteNull();
					else
						property.Value.WriteTo(writer, converters);
				}
			}

			writer.WriteEndObject();
		}

		#region IEnumerable<KeyValuePair<string,RavenJToken>> Members

		public IEnumerator<KeyValuePair<string, RavenJToken>> GetEnumerator()
		{
			return Properties.GetEnumerator();
		}

		#endregion

		#region IEnumerable Members

		IEnumerator IEnumerable.GetEnumerator()
		{
			return GetEnumerator();
		}

		#endregion

		public void Add(string propName, RavenJToken token)
		{
			Properties.Add(propName, token);
		}

		internal override void AddForCloning(string key, RavenJToken token)
		{
			Properties[key] = token;
		}

		public bool Remove(string propName)
		{
			return Properties.Remove(propName);
		}

		public bool ContainsKey(string key)
		{
			return Properties.ContainsKey(key);
		}

		public bool TryGetValue(string name, out RavenJToken value)
		{
			return Properties.TryGetValue(name, out value);	
		}

		public override RavenJToken CreateSnapshot()
		{
			return new RavenJObject(Properties.CreateSnapshot());
		}

		public override void EnsureSnapshot()
		{
			Properties.EnsureSnapshot();
		}

		public override IEnumerable<RavenJToken> Values()
		{
			return Properties.Values;
		}

		public override IEnumerable<T> Values<T>()
		{
			return Properties.Values.Convert<T>();
		}
	}
}

File: Raven.Client.Java/src/main/java/net/ravendb/abstractions/json/linq/RavenJObject.java
package net.ravendb.abstractions.json.linq;

import java.io.IOException;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

import net.ravendb.abstractions.basic.Reference;
import net.ravendb.abstractions.exceptions.JsonReaderException;
import net.ravendb.abstractions.exceptions.JsonWriterException;
import net.ravendb.abstractions.extensions.JsonExtensions;
import net.ravendb.abstractions.json.RavenJsonTextReader;

import org.apache.commons.lang.StringUtils;
import org.codehaus.jackson.JsonGenerator;
import org.codehaus.jackson.JsonParser;
import org.codehaus.jackson.JsonToken;
import org.codehaus.jackson.map.ObjectMapper;

import com.google.common.base.Defaults;


public class RavenJObject extends RavenJToken implements Iterable<Entry<String, RavenJToken>> {

  private final Comparator<String> comparer;
  /**
   * Creates a {@link RavenJObject} from an object.
   * @param o The object that will be used to create {@link RavenJObject}
   * @return A {@link RavenJObject} with the values of the specified object.
   */
  public static RavenJObject fromObject(Object o) {
    return fromObject(o, JsonExtensions.createDefaultJsonSerializer());
  }

  /**
   * Creates a {@link RavenJObject} from an object.
   * @param o The object that will be used to create {@link RavenJObject}
   * @param objectMapper The {@link ObjectMapper} that will be used to read the object.
   * @return {@link RavenJObject} with the values of the specified object.
   */
  public static RavenJObject fromObject(Object o, ObjectMapper objectMapper) {
    RavenJToken token = fromObjectInternal(o, objectMapper);

    if (token != null && token.getType() != JTokenType.OBJECT)
      throw new IllegalArgumentException("Object serialized to " + token.getType()
        + ". RavenJObject instance expected.");

    return (RavenJObject) token;
  }

  public static RavenJObject load(JsonParser parser) {
    try {
      if (parser.getCurrentToken() == null) {
        if (parser.nextToken() == null) {
          throw new JsonReaderException("Error reading RavenJToken from JsonParser");
        }
      }
      if (parser.getCurrentToken() != JsonToken.START_OBJECT) {
        throw new JsonReaderException(
          "Error reading RavenJObject from JsonParser. Current JsonReader item is not an object: "
            + parser.getCurrentToken());
      }
      if (parser.nextToken() == null) {
        throw new JsonReaderException("Unexpected end of json object");
      }

      String propName = null;
      RavenJObject o = new RavenJObject();
      do {

        switch (parser.getCurrentToken()) {
          case FIELD_NAME:
            propName = parser.getText();
            break;
          case END_OBJECT:
            return o;
          case START_OBJECT:
            if (StringUtils.isNotEmpty(propName)) {
              RavenJObject val = RavenJObject.load(parser);
              o.set(propName, val);
              propName = null;
            } else {
              throw new JsonReaderException("The JsonReader should not be on a token of type "
                + parser.getCurrentToken());
            }
            break;
          case START_ARRAY:
            if (StringUtils.isNotEmpty(propName)) {
              RavenJArray val = RavenJArray.load(parser);
              o.set(propName, val);
              propName = null;
            } else {
              throw new JsonReaderException("The JsonReader should not be on a token of type "
                + parser.getCurrentToken());
            }
            break;
          default:
            if (StringUtils.isNotEmpty(propName)) {
              RavenJValue val = (RavenJValue) RavenJToken.load(parser);
              o.set(propName, val);
              propName = null;
            } else {
              throw new JsonReaderException("The JsonReader should not be on a token of type "
                + parser.getCurrentToken());
            }
            break;

        }
      } while (parser.nextToken() != null);

      throw new JsonReaderException("Error reading RavenJObject from JsonReader.");

    } catch (IOException e) {
      throw new JsonWriterException(e.getMessage(), e);
    }
  }

  /**
   * Loads {@link RavenJObject} from a string that contains JSON.
   * @param json A {@link String} that contains JSON.
   * @return A {@link RavenJObject} populated from the string that contains JSON.
   */
  public static RavenJObject parse(String json) {
    try {
      JsonParser jsonParser = new RavenJsonTextReader().createJsonParser(json);
      return load(jsonParser);
    } catch (IOException e) {
      throw new JsonReaderException(e.getMessage(), e);
    }
  }

  private DictionaryWithParentSnapshot properties;

  public RavenJObject() {
    this((Comparator<String>)null);
  }

  public RavenJObject(Comparator<String> comparator) {
    this.comparer = comparator;
    this.properties = new DictionaryWithParentSnapshot(comparator);
  }


  public RavenJObject(DictionaryWithParentSnapshot snapshot) {
    properties = snapshot;
    this.comparer = null;
  }

  public RavenJObject(RavenJObject other) {
    properties = new DictionaryWithParentSnapshot(other.comparer);
    for (Map.Entry<String, RavenJToken> kv : other.getProperties().entrySet()) {
      properties.put(kv.getKey(), kv.getValue());
    }
    this.comparer = other.comparer;
  }

  public void add(String propertyName, Object value) {
    properties.put(propertyName, RavenJToken.fromObject(value));
  }

  public void add(String propertyName, RavenJToken token) {
    properties.put(propertyName, token);
  }

  @Override
  protected void addForCloning(String key, RavenJToken token) {
    properties.put(key, token);
  }

  @Override
  public RavenJObject cloneToken() {
    return (RavenJObject) cloneTokenImpl(new RavenJObject());
  }

  public boolean containsKey(String key) {
    return properties.containsKey(key);
  }

  @Override
  public RavenJObject createSnapshot() {
    return new RavenJObject(properties.createSnapshot());
  }

  @Override
  public void ensureCannotBeChangeAndEnableShapshotting() {
    properties.ensureSnapshot();
  }

  public void ensureSnapshot(String msg) {
    properties.ensureSnapshot(msg);
  }

  public RavenJToken get(String propertyName) {
    return properties.get(propertyName);
  }

  public int getCount() {
    return properties.size();
  }

  public Set<String> getKeys() {
    return properties.keySet();
  }

  /**
   * @return the properties
   */
  public DictionaryWithParentSnapshot getProperties() {
    return properties;
  }

  @Override
  public JTokenType getType() {
    return JTokenType.OBJECT;
  }

  @Override
  public boolean isSnapshot() {
    return properties.isSnapshot();
  }

  public RavenJObject withCaseInsensitivePropertyNames() {
    DictionaryWithParentSnapshot props = new DictionaryWithParentSnapshot(String.CASE_INSENSITIVE_ORDER);
    for (Map.Entry<String, RavenJToken> property: properties) {
      props.put(property.getKey(), property.getValue());
    }
    return new RavenJObject(props);
  }

  public boolean remove(String propertyName) {
    return properties.remove(propertyName) != null;
  }

  public void set(String propertyName, RavenJToken value) {
    properties.put(propertyName, value);
  }

  /**
   * Writes this token to a {@link JsonGenerator}
   * @param writer A {@link JsonGenerator} into which this method will write.
   */
  @Override
  public void writeTo(JsonGenerator writer) {
    try {
      writer.writeStartObject();

      if (properties != null) {
        for (String key : properties.keySet()) {
          writer.writeFieldName(key);
          RavenJToken value = properties.get(key);
          if (value == null) {
            writer.writeNull();
          } else {
            value.writeTo(writer);
          }
        }
      }
      writer.writeEndObject();
    } catch (IOException e) {
      throw new JsonWriterException(e.getMessage(), e);
    }

  }

  /* (non-Javadoc)
   * @see raven.client.json.RavenJToken#value(java.lang.Class, java.lang.String)
   */
  @SuppressWarnings("null")
  @Override
  public <T> T value(Class<T> clazz, String key) {
    if (!containsKey(key)) {
      return Defaults.defaultValue(clazz);
    }
    RavenJToken ravenJToken = get(key);
    if (ravenJToken != null) {
      return Extensions.convert(clazz, ravenJToken);
    }
    throw new IllegalArgumentException("Unsupported conversion. From:" + ravenJToken.getType() + " to "
      + clazz.getCanonicalName());
  }

  @Override
  public Iterable<RavenJToken> values() {
    return properties.values();
  }

  @Override
  public <T> List<T> values(Class<T> clazz) {
    return Extensions.convert(clazz, properties.values());
  }

  @Override
  public Iterator<Entry<String, RavenJToken>> iterator() {
    return properties.iterator();
  }

  public boolean tryGetValue(String name, Reference<RavenJToken> value) {
    return properties.tryGetValue(name, value);
  }

}
